File: java/0020/Main.java
import java.util.*;

class Solution {
  public boolean isValid(String s) {
    Stack<Character> stack = new Stack<>();

    HashMap<Character, Character> dict = new HashMap<>();
    dict.put(')', '(');
    dict.put(']', '[');
    dict.put('}', '{');

    for (char el : s.toCharArray()) {
      if (dict.containsKey(el)) {
        if (stack.isEmpty()) return false;
        if (!dict.get(el).equals(stack.pop())) return false;
      } else {
        stack.push(el);
      }
    }
    
    return stack.isEmpty();
  }
}

File: py/0020/main.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        pairs = {
            ')': '(',
            '}': '{',
            ']': '[',
        }

        for item in s:
            if item in pairs:
                if not stack or pairs[item] != stack.pop():
                    return False
            else:
                stack.append(item)
        return not stack
